geld.lua

```diff
@@ -0,0 +1,133 @@
+-- Gelds animals by default. Optional args: -toggle or -ungeld
+-- Written by Josh Cooper(cppcooper) on 2019-12-10, last modified: 2020-02-23
+utils ={}
+utils = require('utils')
+local validArgs = utils.invert({
+    'unit',
+    'toggle',
+    'ungeld',
+    'help',
+    'find'
+
+})
+local args = utils.processArgs({...}, validArgs)
+local help = [====[
+
+geld
+====
+Geld allows the user to geld and ungeld animals.
+
+Valid options:
+    unit <id> - Performs action on the provided unit id, this is optional.
+                If this argument is not given, the highlighted unit is
+                used instead.
+
+    toggle    - Changes the geld status to the opposite of its current state
+    ungeld    - Sets the geld status to false
+    help      - Shows this help information
+
+]====]
+
+unit=nil
+
+if args.unit then
+    id=tonumber(args.unit)
+    if id then
+        for _,unit_ in pairs (df.global.world.units.active) do
+            if unit.id == id then
+                unit=unit_
+            end
+        end
+    else
+        qerror("Invalid id provided.")
+    end
+else
+    unit = dfhack.gui.getSelectedUnit()
+end
+
+if not unit then
+    qerror("Invalid unit selection.")
+end
+
+oldstate = unit.flags3.gelded
+
+if unit.sex == 0 then
+    qerror("cannot geld female animals")
+    do return end
+end
+
+function exists(thing)
+    if thing then return true else return false end
+end
+
+function FindBodyPart(unit,newstate)
+    bfound=false
+    for i,wound in ipairs(unit.body.wounds) do
+        for j,part in ipairs(wound.parts) do
+            if type(unit.body.wounds[i].parts[j].flags2.gelded) ~= "nil" then
```

Comment:
flags2 appears to be of type wound_damage_flags2, which always has a gelded field present, so it can never be nil: https://github.com/DFHack/df-structures/blob/2d67b03a1c5c701e65fb483b9d6c1d3067f990cb/df.units.xml#L1344-L1348
I think this should just check for parts where gelded is different from `newstate` and set them to `newstate`, but I might be missing something

Reply:
I'll have to double check. I know there were possibly a redundancy in there, but otherwise I know I was very careful to manage the geld state appropriately to avoid creating edge cases.